### ISSUE TITLE
Bug PresentLink when combined with virtual link in some way

### ISSUE THREAD

**Issue by ngeiswei** [opencog/atomspace#2231]
The following 

```scheme
;; KB

(EvaluationLink
  (PredicateNode "P")
  (LambdaLink
    (ImplicationLink
      (VariableNode "$Z")
      (InheritanceLink
        (VariableNode "$X")
        (VariableNode "$Y")
      )
    )
  )
)

(EvaluationLink
  (PredicateNode "P")
  (LambdaLink
    (ImplicationLink
      (VariableNode "$X")
      (VariableNode "$Y")
    )
  )
)

;; Query

(define (dummy x) (stv 1 1))

(define query
  (BindLink
    (AndLink
      (EvaluationLink
        (GroundedPredicateNode "scm: dummy")
        (EvaluationLink
          (PredicateNode "P")
          (LambdaLink
            (ImplicationLink
              (InheritanceLink
                (VariableNode "$X")
                (VariableNode "$Y")
              )
              (VariableNode "$Z")
            )
          )
        )
      )
      (PresentLink
        (EvaluationLink
          (PredicateNode "P")
          (LambdaLink
            (ImplicationLink
              (InheritanceLink
                (VariableNode "$X")
                (VariableNode "$Y")
              )
              (VariableNode "$Z")
            )
          )
        )
      )
    )
    (Concept "OK")
  )
)

(cog-execute! query)
```

raises an exception

```scheme
ERROR: In procedure opencog-extension:
Throw to key `C++-EXCEPTION' with args `("cog-execute!" "Not implemented! Need to implement a stack, here. (/home/nilg/OpenCog/atomspace/opencog/query/DefaultPatternMatchCB.cc:265)
Function args:
...
```

If `Present` is removed from the query, then it works.

**Comment by ngeiswei** [comment]
I have created a unit test for it https://github.com/ngeiswei/atomspace/commits/fix-present-bug and I am investing the problem.

**Comment by linas** [comment]
That error message might be misleading.  The BindLink has zero variables in it.  The PresentLink is wrapping something that is effectively a constant: there are zero free variables inside that PresentLink, so that term will always be in the atomspace, by definition, so the PresentLink is trivially satisfied.  The pattern matcher should have printed a warning about constant clauses, and removed that clause.

That leaves you with just the GPN, which is also wrapping a constant. The GPN should print the constant, and be done.

I conclude the constant-clause-remove code should have noticed that the PresentLink is wrapping a constant, and removed it. The result will never trigger that assert.

**Comment by linas** [comment]
Hmm Seems that it was correctly identified as a constant clause, and it was removed. This removal no longer prints any warning; its silent. I guess that's OK.

I don't understand why  #2233 fixes this, because the result should have been .. hmm. Odd.

**Comment by linas** [comment]
... yet, somehow, the constant was not removed, the pattern got split into two pieces, each piece got evaluated, which is why  #2233 fixes it.  But the split should not have happened in the first place.... looking into that.

**Comment by ngeiswei** [comment]
Yes, the split happened because `PresentLink` flags the clause as quoted, however I do not completely understand the C++ `PatternLink` class so I could not fix it under that angle, and #2233 is a worthwhile fix beyond that bug, so I went for it.

**Comment by linas** [comment]
OK so #2235 fixes the constant-terms bug. If this had been done first, the fix for for #2233 would never have been found :-(

**Comment by linas** [comment]
>  do not completely understand the C++ PatternLink class 

The quoted/unquoted/fixed/virtual/mandatory/optional stuff is a ball of confusion; its the legacy of it being old code. I tried real hard to untangle some of this, a few weeks(? months ?) ago, and was partially but not completely successful. I gave up after I could not find any further simplifications.  In short, each of these distinctions between clause-types need to be made for various reasons...
